### ISSUE TITLE
enable indexed search of species interaction data by bionomia id

### ISSUE THREAD

**Issue by jhpoelen** [bionomia/bionomia#112]
Use case:

1. visit https://globalbioticinteractions.org 
2. enter biononia id (e.g., wikidata/orcid) in the "according to" field
3. inspect all species interactions linked to a specific person 

In order to build an index to link species interaction records by their occurrence id to bionomia, access to all of bionomia links are needed. However, aside from scraping the bionomia's web apis or html pages, I was not able find a way to download all of bionomia. 

How do you imagine others can re-use bionomia's ever growing collection of links between specimen (records) and the humans that contributed to them?

**Comment by dshorthouse** [comment]
Good question. Personal needs for privacy are layered in this story as well. Nonetheless, the question is also, "In what form should that download take?" I'd be reluctant to make a "download all of Bionomia" because it's really about being a broker of links between entities like `occurrenceID`, person, and the action they executed (collected, identified) whereby all the GBIF records get dumped and refreshed every few weeks. As you know, these `occurrenceID` tend to evaporate. Versioning is tricky in such a model, especially since the fields captured from GBIF are only those that effectuate the linking. That said, there are dataset-level downloads as Frictionless Data, eg https://bionomia.net/dataset/84ab7b76-f762-11e1-a439-00145eb45e9a. The intent here is for a consumable structure that collections managers might be able to use because it's more relational than that of a DwC-A. So... what would/could a "download all of Bionomia" look like?

**Comment by jhpoelen** [comment]
Thanks for your prompt reply.

For the GloBI use case the following table might work:

bionomiaPersonId | bionomiaPersonName | occurrenceId | catalogNumber | collectionCode | institutionId | gbif:datasetId | scientificName | lat | long | eventDate | bionomia version | bionomia datetime | 
--- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | 
orcid/wd | Jane Doe (human readable) | uuid/uri | 1234 | ABCD | DEF | Homo sapiens | 32.2 | 41.8 | 2021-01-01 | version of bionomia | date/time versioned was created | 

So, a minimal set of identifying information for records and the humans connected. I am sure that roles of the humans can be documented in more detail (e.g., collectedBy, identifiedBy), but I imagine that would be relatively easy to add if needed.

**Comment by jhpoelen** [comment]
The individual data packages look nice, but hard to aggregate. 

e.g., 
```
cat package1.zip package2.zip > combined.zip
```

does not produce a happy zip file.

Perhaps create a "frictionless" data package for all of bionomia?

**Comment by dshorthouse** [comment]
@jhpoelen Thanks for the example. Yes, I'd also include `recordedBy` and `identifiedBy` along with a column for the action the person executed. So, the result would be quite an explosion of entries (because more than one person was a listed as a collector or an identifier per occurrence), but would nonetheless indicate with a bit more precision about _why_ a person is connected to a particular occurrence record.

**Comment by dshorthouse** [comment]
> Perhaps create a "frictionless" data package for all of bionomia?

Yes, that's possible and I'll see how I might schedule/version such a thing with every reharvest/refresh of GBIF. It will be very large.

**Comment by jhpoelen** [comment]
@dshorthouse I like big things, especially when you can stream them. How big is big in this case? 

One way to do publish bionomia whole data index might be to have the data package metadata and the referenced data separately in a stream-able friendly format, like individual tsv.gz files combined with a single package json. Perhaps as an unpacked zenodo publication. 

Zip files can be streamed also, but the tooling and ordering of meta-data (package json would have to appear before the data) can be a bit of a pain.

**Comment by dshorthouse** [comment]
> @dshorthouse I like big things, especially when you can stream them. How big is big in this case?

Not sure I can stream them because there's logic at play in the middle layer. I reworked some bits to better produce an "all Bionimia" Frictionless dataset, tried it out locally, & we're looking at a file approx. 1GB zipped.

**Comment by jhpoelen** [comment]
@dshorthouse thanks for the quick prototyping efforts.

your suggestion to

> [...]  produce an "all Bionimia" Frictionless dataset, tried it out locally, & we're looking at a file approx. 1GB zipped.

looks like a great start. Is there a way you can share this file, so I can work on a first pass at a Bionomia <> GloBI integration ?

**Comment by jhpoelen** [comment]
PS. the 1GB size looks small enough to be used without too many issues, especially when the content can be accessed in streaming fashion (e.g., line-by-line after independently accessing some schema meta-data)

Looking forward!

**Comment by dshorthouse** [comment]
Here's a first crack at it. Performance in generating it is not v. good because it's not a straight dump. I have more work to do to operationalize this. I cannot realistically produce this more frequently than once a month or so & the trigger is manual.

https://doi.org/10.5281/zenodo.4764044

**Comment by jhpoelen** [comment]
@dshorthouse Thanks for putting this re-usable whole binomia dataset together. I am hoping to find the time to integrate this data to enable searches for interaction record by orcid ids the folks providing them. https://github.com/globalbioticinteractions/globalbioticinteractions/issues/658

**Comment by jhpoelen** [comment]
One suggestion/request for future publication: 

instead of publishing a zip file, suggest to publish independent bz2 or gzip compressed data files along with uncompressed data package info. This way, you can more easily directly stream/process the data from the data publication rather than having to first download a zip file, extract a package meta-data only to find out that the data is unable to be processed by the recipient.

**Comment by dshorthouse** [comment]
> instead of publishing a zip file, suggest to publish independent bz2 or gzip compressed data files along with uncompressed data package info.

I’d assuming something like that is possible with Frictionless Data but I see nothing in the specs/examples. The closest I see is to split each resource into multiple csv files such that the path becomes an array, each of which may be held remote to the descriptor. Am guessing here that the problem is when you include the schema of the resource as I’ve done here, its format (= csv), and its media type (= text/csv), how does one  _additionally_ say that these are gzipped such that any validator knows what to do with the resource prior to applying the schema?

**Comment by dshorthouse** [comment]
Just found this. Evidently it’s a known problem and likely request but not yet formally part of the spec: https://specs.frictionlessdata.io/patterns/#meta

**Comment by jhpoelen** [comment]
@dshorthouse Thanks for having a look into this. Appears that in conveniently co-locating data and related schema/ meta-data, the ability to separate them physically, but not logically, has been given less attention. 

I spent some time to derive a terse Bionomia dataset that would serve my purposes, has explicit provenance, is location-agnostic and can be verified. It is a bit of a prototype, so lots of room for improvement.

For your review:

Poelen, Jorrit H. (2021). attributions.tsv.gz: a Terse Biodiversity Data Attributions Dataset derived from Bionomia (Version 0.0.3) [Data set]. Zenodo. http://doi.org/10.5281/zenodo.4779069

Note that this publication can be integrated with other available Preston-provenance-enabled datasets like DataONE, BHL, GBIF, iDigBio, OBIS etc. to help answer questions like: How many Meise Botanical Garden or Arctos specimen occurrence ids have been attributed to wikipedia or orcid ids via Bionomia (see https://github.com/bio-guoda/preston/issues/113)?

fyi @mielliott @qgroom

**Comment by jhpoelen** [comment]
fyi @cboettiger (an example of machine-readable provenance layered on existing and derived dataset published via Zenodo).

**Comment by dshorthouse** [comment]
That's great @jhpoelen. A point of clarification.
 
You provide the following entitled, "attributions":

```
http://arctos.database.museum/guid/MSB:Host:15730?seid=2593587    identifiedBy    https://orcid.org/0000-0003-4738-5967
http://arctos.database.museum/guid/MSB:Host:15731?seid=2593588    identifiedBy    https://orcid.org/0000-0003-4738-5967
LD:General:1001542    identifiedBy    http://www.wikidata.org/entity/Q86532
NRM:NRM-Fish:16167    identifiedBy    http://www.wikidata.org/entity/Q446239
```
I'd call those associations or a naked triple. Missing from this is who made the association - the attribution - as part of the provenance chain. In the first example above it was https://orcid.org/0000-0002-1970-7044 who made the attribution at 2019-10-31 21:28:43.

**Comment by jhpoelen** [comment]
@dshorthouse Thanks for your prompt reply and review.

Yes, the attributions.tsv.gz contain the "naked" triples, or attribution shortcuts, attributions.tsv.gz, captures the (naked) attribution of records. An additional resource (e.g., attributions_of_attributions.tsv.gz) can be provided to relate the scribe (via their orcid) to the claimed attribution. This can still be derived from the provided resource. I imagine a provenance ontology like notation can be used to describe the process/activity of the scribe, in which a person (the scribe) uses a record (identified by occurrenceId, gbifID etc.) and additional information (e.g., verbatim collector information) to generate a relation between a record a person via an orchid/wikidata identifier.  

For the time being, the naked attribution is sufficient for my purposes, assuming that the attributions_of_attributions.tsv.gz can be generated when needed or looked up via Bionomia search / website. 

PS. Is there an easy way to link into Bionomia by occurrenceId along with their orchid / wikidata to show related claims?

**Comment by dshorthouse** [comment]
> PS. Is there an easy way to link into Bionomia by occurrenceId along with their orchid / wikidata to show related claims?

By `gbifID` (not `occurrence_id` because these are not unique nor always present), but how are you defining a claim or relatedness? Do you mean:

- Given an ORCID or a wikidata Q number, show me all that person's attributed occurrence records?
- Given an ORCID or a wikidata Q number, show me all that person's occurrence records attributed by other people?
- Given an ORCID, show me the occurrence records they claimed as their own?
- Differentiate any of the above according to either collected or identified?

**Comment by jhpoelen** [comment]
Desired link would provide a Bionomia web context that helps answer the question:

Who claimed that occurrenceId X was identifiedBy/recordedBy by preston Y (as identified by orcid id / wikidata id)?

I prefer using occurrenceId because these are the identifiers most likely present in original data archives as provided by the institutions. I realize that occurrenceIds are not always provided, and I am ok with having an incomplete, or subset, of attributions in the derived resource attributions.tsv.gz for now.

**Comment by jhpoelen** [comment]
Btw - some statistics about the "naked" triples from :

```
$ preston cat --remote https://zenodo.org/record/4779069/files hash://sha256/ce5ef67da6ad41e88ae3df60a1c5329273cf1c33a34955153d2d9a6022b7e0d5 | gunzip | cut -f2 | sort | uniq -c |sort -nr
11415597 recordedBy
5215915 identifiedBy
```

report ~11M recordedBy and ~5M identifiedBy attributions. This is a subset of the total amount of attributions, because only attribution claims with occurrenceIds are included.

**Comment by dshorthouse** [comment]
> I prefer using occurrenceId because these are the identifiers most likely present in original data archives as provided by the institutions.

True, but none are stable across the board & there's nothing in `occurrenceID` that always gives me a consistent & known landing page for users of Bionomia to go look at something prior to making an attribution. I'll see what impact indexing by `occurrenceID` might impose but cannot now guarantee that I can do this well.

**Comment by dshorthouse** [comment]
> This is a subset of the total amount of attributions, because only attribution claims with occurrenceIds are included.

And even then by privacy policy, including only publicly-available ORCID profiles but all Wikidata Q number-based ones.

**Comment by jhpoelen** [comment]
Yes, occurrenceIds are provided as is and their uniqueness/stability (just like gbifIDs, orcid, wikidata) depend on the data management practices of those curating the related data platforms. 

Perhaps the most pragmatic way is to link into Bionomia by orcid / wikidata id . This way, you at least narrow the search space to the related person and the platform that curates the attribution claims.

With this, the provided link would answer the question:

Which attribution claims are available for person X (by orcid/wikidata id) and who made these attribution claims? 

Assuming that a linking mechanism into this (dynamic) Bionomia web context is available, what would be your preferred link into Bionimia?

**Comment by dshorthouse** [comment]
Perhaps then https://bionomia.net/0000-0002-1970-7044 or https://bionomia.net/Q6077932 is good enough for now. There's JSON-LD content negotiation on that URL as well. While I have not yet figured out how to represent who made attributions in JSON-LD, I'll eventually get this figured out.

**Comment by dshorthouse** [comment]
Not sure what I was thinking, I do in fact show the attributor when known with an ORCID. Approx. 1.4M attributions come from source via `identifiedByID` or `recordedByID` & so have no idea who made it so those instances do not have a `creator`. Otherwise, they do:

```
curl -LH "Accept: application/ld+json" "https://bionomia.net/Q50655527"
```
